app/post.go

```diff
@@ -1174,3 +1174,131 @@ func (a *App) MaxPostSize() int {
 
 	return maxPostSize
 }
+
+func (a *App) countMentionsFromPost(user *model.User, post *model.Post) (int, *model.AppError) {
+	channel, err := a.GetChannel(post.ChannelId)
+	if err != nil {
+		return 0, err
+	}
+
+	if channel.Type == model.CHANNEL_DIRECT {
+		// Return -1 to indicate that the mention count should be equal to the post count
+		return -1, nil
+	}
+
+	channelMember, err := a.GetChannelMember(channel.Id, user.Id)
+	if err != nil {
+		return 0, err
+	}
+
+	keywords := addMentionKeywordsForUser(
+		map[string][]string{},
+		user,
+		channelMember.NotifyProps,
+		&model.Status{Status: model.STATUS_ONLINE}, // Assume the user is online since they would've triggered this
+		true, // Assume channel mentions are always allowed for simplicity
+	)
+	commentMentions := user.NotifyProps[model.COMMENTS_NOTIFY_PROP]
+	checkForCommentMentions := commentMentions == model.COMMENTS_NOTIFY_ROOT || commentMentions == model.COMMENTS_NOTIFY_ANY
+
+	// A mapping of thread root IDs to whether or not a post in that thread mentions the user
+	mentionedByThread := make(map[string]bool)
+
+	isCommentMention := func(post *model.Post, otherPosts map[string]*model.Post) bool {
```

Comment:
are we making a closure over some variable of the method? if not, we should move this out, as it is way too big to have it inside. if yes, I'd recomend to move it out and return a function given the closured variable.
I really like defining small helper functions inside other functions, but only if they are tiny. It improves readability, having a big one does the opposite as you have to figure out if you are on one or the other scope.